File: NoTomorrow/Features/Workout/ExerciseEquipment.swift
import Foundation

extension ExerciseLibrary {
    /// Equipment filter chips of the exercise picker, mapped onto free-exercise-db `equipment` values: the EZ bar
    /// counts as a barbell, and medicine / exercise balls, foam rollers, "other" and no equipment at all are Other.
    enum Equipment: String, CaseIterable, Identifiable {
        case all, barbell, dumbbell, machine, cable, bodyweight, kettlebell, band, other
        var id: String { rawValue }

        var titleKey: String {
            switch self {
            case .all: "equipment.all"
            case .barbell: "equipment.barbell"
            case .dumbbell: "equipment.dumbbell"
            case .machine: "equipment.machine"
            case .cable: "equipment.cable"
            case .bodyweight: "equipment.body_only"
            case .kettlebell: "equipment.kettlebells"
            case .band: "equipment.bands"
            case .other: "equipment.other"
            }
        }

        /// The chip an exercise's raw `equipment` value falls under (never `.all`).
        static func of(_ equipment: String?) -> Equipment {
            switch equipment?.lowercased() {
            case "barbell", "e-z curl bar": .barbell
            case "dumbbell": .dumbbell
            case "machine": .machine
            case "cable": .cable
            case "body only": .bodyweight
            case "kettlebells": .kettlebell
            case "bands": .band
            default: .other
            }
        }

        func matches(_ equipment: String?) -> Bool {
            self == .all || Self.of(equipment) == self
        }

        /// The raw value a custom exercise created under this chip gets, so it stays in the filtered list.
        var representative: String? {
            switch self {
            case .all: nil
            case .barbell: "barbell"
            case .dumbbell: "dumbbell"
            case .machine: "machine"
            case .cable: "cable"
            case .bodyweight: "body only"
            case .kettlebell: "kettlebells"
            case .band: "bands"
            case .other: "other"
            }
        }
    }
}

File: android/app/src/main/java/app/notomorrow/feature/workout/ExerciseEquipment.kt
package app.notomorrow.feature.workout

import androidx.annotation.StringRes
import app.notomorrow.data.entity.ExerciseEntity
import app.notomorrow.util.S

/**
 * `ExerciseLibrary.Equipment` — the exercise picker's equipment chips, mapped onto
 * free-exercise-db `equipment` values: the EZ bar counts as a barbell, and medicine / exercise
 * balls, foam rollers, "other" and no equipment at all are [Other].
 */
enum class ExerciseEquipment(val raw: String) {
    All("all"),
    Barbell("barbell"),
    Dumbbell("dumbbell"),
    Machine("machine"),
    Cable("cable"),
    Bodyweight("bodyweight"),
    Kettlebell("kettlebell"),
    Band("band"),
    Other("other");

    /** The chip title — the equipment names the exercise detail already uses. */
    @get:StringRes
    val titleRes: Int
        get() = when (this) {
            All -> S.equipment_all
            Barbell -> S.equipment_barbell
            Dumbbell -> S.equipment_dumbbell
            Machine -> S.equipment_machine
            Cable -> S.equipment_cable
            Bodyweight -> S.equipment_body_only
            Kettlebell -> S.equipment_kettlebells
            Band -> S.equipment_bands
            Other -> S.equipment_other
        }

    /** The raw value a custom exercise created under this chip gets, so it stays in the filtered list. */
    val representative: String?
        get() = when (this) {
            All -> null
            Barbell -> "barbell"
            Dumbbell -> "dumbbell"
            Machine -> "machine"
            Cable -> "cable"
            Bodyweight -> "body only"
            Kettlebell -> "kettlebells"
            Band -> "bands"
            Other -> "other"
        }

    fun matches(equipment: String?): Boolean = this == All || of(equipment) == this

    companion object {
        /** The chip an exercise's raw `equipment` value falls under (never [All]). */
        fun of(equipment: String?): ExerciseEquipment = when (equipment?.lowercase()) {
            "barbell", "e-z curl bar" -> Barbell
            "dumbbell" -> Dumbbell
            "machine" -> Machine
            "cable" -> Cable
            "body only" -> Bodyweight
            "kettlebells" -> Kettlebell
            "bands" -> Band
            else -> Other
        }

        /** The picker's equipment filter, applied after the search and muscle chip. */
        fun filter(exercises: List<ExerciseEntity>, equipment: ExerciseEquipment): List<ExerciseEntity> =
            if (equipment == All) exercises else exercises.filter { equipment.matches(it.equipment) }
    }
}
